### PR TITLE
[CORE24-115] task(scim): Integrate SCIM provisioning for Okta

### PR DESCRIPTION
**Pull request by dfigueroa87**
Jira ticket: JIRA-CORE24-115

## What have you implemented?

Minimal SCIM server following SCIM specification for integrating with Okta.

Also integrated `zod` for validation.

For Authorisation I defined and used an API token. In the future we may want to switch to OAuth2

Added `dotenv` for local development, to easily set the values of env vars. You just need to create a `.env` file in the root of the project and add the variables there. E.g.:
```
OKTA_SCIM_API_TOKEN=test-token
```

## How to test it?

1st test every flow defined in [Okta docs](https://developer.okta.com/docs/reference/scim/scim-20/)
Then integrate with Okta and test from there.

## Extra/next steps

- [ ] Integrate in Okta
- [x] Add OKTA_SCIM_API_TOKEN environment variable with the API token configured in Okta

### CODE REVIEW

File: apps/core-api/src/controllers/scim.controller.ts [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+  };
+};
+
+const errorHandlerMiddleware = (
+  error: Error,
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  // TODO: log error
```
Comment (internetti) [inline apps/core-api/src/controllers/scim.controller.ts:34]: is this todo waiting for something specific? is this blocked?

Comment (dfigueroa87) [reply apps/core-api/src/controllers/scim.controller.ts:34]: Thanks, I totally forgot about this. 
I was first thinking of including a logging library as part of these changes, but was a bit wary about how the PR was getting bigger.
For now I'll just add a `console.error()` here, and deal with the logging lib in a new PR, if that's fine with everyone

File: apps/core-api/src/controllers/scim.controller.ts [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+const createScimErrorResponse = (status: number, detail: string) => {
+  return {
+    schemas: ['urn:ietf:params:scim:api:messages:2.0:Error'],
+    status,
+    detail,
+  };
+};
+
+const errorHandlerMiddleware = (
```
Comment (internetti) [inline apps/core-api/src/controllers/scim.controller.ts:28]: I would move some of these things into different files/folders, like the router, and this middleware

Comment (internetti) [reply apps/core-api/src/controllers/scim.controller.ts:28]: just saw that you probably put everything here because the welcome router did, but I'd still split things up. but we can postpone this discussion

File: apps/core-api/src/controllers/scim.controller.ts [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+  }
+
+  next();
+};
+
+const router = Router();
+router.use(authorise);
+router.use(errorHandlerMiddleware);
+
+router.post('/Users', validateCreateUserRequest, async (req, res, next) => {
```
Comment (internetti) [inline apps/core-api/src/controllers/scim.controller.ts:113]: style comment, aren't urls usually lowercase?

File: apps/core-api/src/controllers/scim.controller.ts (hunk 2 above)
Comment (dfigueroa87) [reply apps/core-api/src/controllers/scim.controller.ts:28]: Yeah, I kind of liked the idea of having the router defined and exported by the controllers. The middlewares can totally go elsewhere. Let me see how to re-organise it better :+1:

File: apps/core-api/src/controllers/scim.controller.ts (hunk 3 above)
Comment (dfigueroa87) [reply apps/core-api/src/controllers/scim.controller.ts:113]: Usually yes, but they are case sensitive. And the SCIM specification says they have to be like this. :shrug: 

For our own defined routes let's definitely go lowercase

File: apps/core-api/src/data-access/user.da.ts
Comment (neb42) [inline apps/core-api/src/data-access/user.da.ts:1]: I have a stylistic preference to rename the functions from `createUser` to `create` and import as `import * as UserDA from './user.da';`. I think it makes it easier to understand where a function comes from when it is used in the service/controller. But that is just my preference.

File: apps/core-api/src/data-access/user.da.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import { db, PostgresError, PostgresErrorCode } from '@nrcno/db';
+
+import { User } from '../models/user.model';
+import { AlreadyExistsError } from '../errors';
+
+export const createUser = async (user: Partial<User>): Promise<User> => {
```
Comment (neb42) [inline apps/core-api/src/data-access/user.da.ts:6]: Why partial?

File: apps/core-api/src/models/user.model.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+interface Email {
+  primary?: boolean;
+  value: string;
+  type?: string;
+}
+
+export class User {
+  id: string;
```
Comment (neb42) [inline apps/core-api/src/models/user.model.ts:8]: I think we should have an internal user id and the okta id stored for the user

File: apps/core-api/src/models/user.model.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
+interface Email {
+  primary?: boolean;
+  value: string;
+  type?: string;
+}
+
+export class User {
```
Comment (neb42) [inline apps/core-api/src/models/user.model.ts:7]: I would use a zod type in place of a model. I don't think the model adds a huge amount of value and the zod type is needed for validation anyway.

File: apps/core-api/src/controllers/scim.controller.ts [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+  };
+};
+
+const errorHandlerMiddleware = (
+  error: Error,
+  req: Request,
+  res: Response,
+  next: NextFunction,
+) => {
+  console.error('SCIM controller', error);
```
Comment (neb42) [inline apps/core-api/src/controllers/scim.controller.ts:34]: We should setup a better logging solution/pattern. Happy to leave this for now, but before any more stories.

File: apps/core-api/src/controllers/scim.controller.ts [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+  next();
+};
+
+const router = Router();
+router.use(authorise);
+router.use(errorHandlerMiddleware);
+
+router.post('/Users', validateCreateUserRequest, async (req, res, next) => {
+  try {
+    const user = await createUser(req.body);
```
Comment (neb42) [inline apps/core-api/src/controllers/scim.controller.ts:115]: `req.body` should be validated with zod before passing to the data layer

File: apps/core-api/src/controllers/scim.controller.ts [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+    next(error);
+  }
+});
+
+router.get('/Users/:id', validateUserIdParam, async (req, res, next) => {
+  try {
+    const user = await getUser(req.params.id);
+    if (user) {
+      const scimUser = mapUserToScimUser(user);
+      res.json(scimUser);
```
Comment (neb42) [inline apps/core-api/src/controllers/scim.controller.ts:136]: I think it's better to explicitly send a status code

File: apps/core-api/src/controllers/scim.controller.ts [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+  }
+});
+
+router.put(
+  '/Users/:id',
+  validateUserIdParam,
+  validateCreateUserRequest,
+  async (req, res, next) => {
+    try {
+      const user = await updateUser(req.params.id, req.body);
```
Comment (neb42) [inline apps/core-api/src/controllers/scim.controller.ts:152]: Same comment about body validation

File: apps/core-api/src/controllers/scim.controller.ts [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+      }
+    } catch (error) {
+      next(error);
+    }
+  },
+);
+
+router.patch('/Users/:id', validateUserIdParam, async (req, res, next) => {
+  try {
+    const patchOperations = req.body.Operations;
```
Comment (neb42) [inline apps/core-api/src/controllers/scim.controller.ts:168]: validate body

File: apps/core-api/src/controllers/scim.controller.ts [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+
+    const update = {};
```
Comment (neb42) [inline apps/core-api/src/controllers/scim.controller.ts:170]: This could do with a type

File: apps/core-api/src/controllers/scim.controller.ts [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+      users = await listUsers(startIndex - 1, count);
+      totalResults = await getUserCount();
+    }
+
+    const scimUsers = users.map(mapUserToScimUser);
+    const listResponse = {
+      schemas: ['urn:ietf:params:scim:api:messages:2.0:ListResponse'],
+      totalResults,
+      Resources: scimUsers,
+      startIndex,
```
Comment (neb42) [inline apps/core-api/src/controllers/scim.controller.ts:244]: When searching start index isn't used, so this should be set to `0` if a search is active

File: apps/core-api/src/controllers/scim.controller.ts (hunk 8 above)
Comment (dfigueroa87) [reply apps/core-api/src/controllers/scim.controller.ts:115]: `validateCreateUserRequest` middleware does that

File: apps/core-api/src/controllers/scim.controller.ts [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+router.patch('/Users/:id', validateUserIdParam, async (req, res, next) => {
+  try {
+    const patchOperations = req.body.Operations;
+
+    const update = {};
+
+    for (const operation of patchOperations) {
+      // Our Okta integration should only send replace operations for the active field
+      if (operation.op === 'replace' && operation.value?.active !== undefined) {
+        Object.assign(update, { active: operation.value.active });
```
[the remark this answers is not in the document]
Comment (dfigueroa87) [reply apps/core-api/src/controllers/scim.controller.ts:175]: According to Okta documentation, user details updates go through the [PUT endpoint](https://developer.okta.com/docs/reference/scim/scim-20/#update-a-specific-user-put).
The [PATCH endpoint](https://developer.okta.com/docs/reference/scim/scim-20/#update-a-specific-user-patch) is only used to activate/deactivate users, and to sync user's password (which I didn't implement since I don't think we'll need that feature)

So in short, to keep things simple I reduced the scope of this endpoint to only deal with the use cases we do need right now.

File: apps/core-api/src/services/user.service.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+  if (name?.familyName) {
+    user.lastName = name.familyName;
+  }
+
+  return user;
+};
+
+export const createUser = async (scimUser: ScimUser): Promise<User> => {
+  const user = mapScimUserToUser(scimUser);
+  user.id = uuidv4();
```
[the remark this answers is not in the document]
Comment (dfigueroa87) [reply apps/core-api/src/services/user.service.ts:38]: The `id` field is meant to be generated by us. `userName` should be sort of a user-friendly id from Okta, and I read some bits about an `externalId` field in the documentation that didn't completely understand, but I can take a better look to see if that's an Okta generated id that we could persist.

File: apps/core-api/src/services/user.service.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+
+  if (name?.givenName) {
+    user.firstName = name.givenName;
+  }
+
+  if (name?.familyName) {
+    user.lastName = name.familyName;
+  }
+
+  return user;
```
[the remark this answers is not in the document]
Comment (dfigueroa87) [reply apps/core-api/src/services/user.service.ts:33]: I generally agree, but when doing so we need to be mindful of the impact of having an attribute vs not having it, as they are different things.

I'll use `undefined` instead of `null` though, since for partial updates we don't want to overwrite possible values stored in the database.

File: apps/core-api/src/controllers/scim.controller.ts (hunk 13 above)
Comment (dfigueroa87) [reply apps/core-api/src/controllers/scim.controller.ts:244]: SCIM specs use 1-based index, and that's the default value set when not used. So I think it's fine to return it like that.

File: apps/core-api/src/data-access/user.da.ts [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+  startIndex: number,
+  count: number,
+): Promise<User[]> => {
+  const rows = await db('users').offset(startIndex).limit(count);
+  return rows.map((row) => new User(row));
+};
+
+export const searchUsers = async (
+  attribute: string,
+  value: string,
```
[the remark this answers is not in the document]
Comment (dfigueroa87) [reply apps/core-api/src/data-access/user.da.ts:52]: I think splitting it into two functions makes it simpler to use and understand (given that for our current use cases we don't need to have both functionalities together). 
But not a strong opinion, I'll merge them into a single function.

File: apps/core-api/src/controllers/scim.controller.ts (hunk 8 above)
Comment (neb42) [reply apps/core-api/src/controllers/scim.controller.ts:115]: That may be validating the body (the most important bit) but `req.body` is still `any`

File: apps/core-api/src/data-access/user.da.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (neb42) [reply apps/core-api/src/data-access/user.da.ts:52]: Happy to leave this split out, but for other features I would merge them together

File: apps/core-api/src/data-access/user.da.ts (hunk 4 above)
Comment (dfigueroa87) [reply apps/core-api/src/data-access/user.da.ts:6]: I think the `id` generation belongs to the service layer. But you're right about the type, this should be `Omit<User, 'createdAt' | 'updatedAt'>

File: libs/db/src/lib/db-utils.ts [hunk 18]
```diff
@@ -0,0 +1,36 @@
+const toCamelCase = (str: string): string => {
+  return str.replace(/([-_][a-z])/g, (group) =>
+    group.toUpperCase().replace('-', '').replace('_', ''),
+  );
+};
+
+const toSnakeCase = (str: string): string => {
+  return str.replace(/([A-Z])/g, '_$1').toLowerCase();
+};
```
[the remark this answers is not in the document]
Comment (dfigueroa87) [reply libs/db/src/lib/db-utils.ts:9]: I wouldn't expect to need this elsewhere, but indeed we can move it when needed (or even decide to use something like lodash instead)

File: apps/core-api/src/controllers/scim.controller.ts (hunk 7 above)
Comment (dfigueroa87) [reply apps/core-api/src/controllers/scim.controller.ts:36]: Agree, this is a temporary solution until we integrate a library and setup logging properly, as mentioned in an earlier comment

File: package.json [hunk 19]
```diff
@@ -20,9 +20,12 @@
     "react": "18.2.0",
     "react-dom": "18.2.0",
     "react-router-dom": "6.11.2",
-    "tslib": "^2.3.0"
+    "tslib": "^2.3.0",
+    "uuid": "^9.0.1",
```
[the remark this answers is not in the document]
Comment (dfigueroa87) [reply package.json:25]: ulid looks good to me. But let's do this later on, I think it's not relevant for user ids anyway

File: apps/core-api/src/services/user.service.ts
Comment (neb42) [inline apps/core-api/src/services/user.service.ts:1]: Also same comment as the store. Naming things as `create` and used as `UserService.create`

File: apps/core-api/src/models/user.model.ts [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+});
+
+export const UserSchema = z.object({
+  id: z.string(),
+  oktaId: z.string().optional(),
+  userName: z.string(),
+  firstName: z.string().optional(),
+  lastName: z.string().optional(),
+  displayName: z.string().optional(),
+  emails: z.array(EmailSchema).optional(),
```
Comment (neb42) [inline apps/core-api/src/models/user.model.ts:16]: Maybe add `.default([])` as parsing an empty array is easier than checking for null

File: apps/core-api/src/data-access/user.store.ts [hunk 21]
```diff
@@ -0,0 +1,73 @@
+import { db, PostgresError, PostgresErrorCode } from '@nrcno/db';
```
Comment (neb42) [inline apps/core-api/src/data-access/user.store.ts:1]: I think the import should be a bit more explicity. Seen as `@nrcno` is an actual npm org, this could be an actual package. Maybe `@nrcno/core-db`

File: apps/core-api/src/data-access/user.store.ts [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+
+import { User, UserSchema } from '../models/user.model';
+import { AlreadyExistsError } from '../errors';
+
+export const create = async (
+  user: Omit<User, 'createdAt' | 'updatedAt'>,
```
Comment (neb42) [inline apps/core-api/src/data-access/user.store.ts:7]: This should be defined as a schema and a type so that we can validate it.

File: apps/core-api/src/services/user.service.ts [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+    lastName: name?.familyName,
+  };
+};
+
+export const create = async (scimUser: ScimUser): Promise<User> => {
+  const user = {
+    ...mapScimUserToUser(scimUser),
+    id: uuidv4(),
+  };
+  return UserStore.create(user as Omit<User, 'createdAt' | 'updatedAt'>);
```
Comment (neb42) [inline apps/core-api/src/services/user.service.ts:28]: I don't like that this we are casting `user` here. I get why but can't think of a clean alternative off the top of my head.

**Review by neb42 — APPROVED** [review]

**Comment by neb42** [comment]
I haven't tested this yet. I'll wait until the full story is done.

File: apps/core-api/src/services/user.service.ts (hunk 23 above)
Comment (dfigueroa87) [reply apps/core-api/src/services/user.service.ts:28]: I agree, don't like it either but also couldn't find another way
